src/libopensc/card-piv.c

```diff
@@ -1974,19 +1971,27 @@ static int piv_general_external_authenticate(sc_card_t *card,
 	 * memcopy the body past the 7C<len> portion
 	 * Transmit
 	 */
-	tmplen = put_tag_and_len(0x82, cypher_text_len, NULL);
+	tmplen = sc_asn1_put_tag(0x82, NULL, cypher_text_len, NULL, 0, NULL);
+	if (tmplen <= 0) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
 
-	tmplen = put_tag_and_len(0x7C, tmplen, &p);
+	r = sc_asn1_put_tag(0x7C, NULL, tmplen, p, output_len, &p);
+	if (r != SC_SUCCESS) {
+		goto err;
+	}
 
 	/* Build the 0x82 TLV and append to the 7C<len> tag */
-	tmplen += put_tag_and_len(0x82, cypher_text_len, &p);
-
-	memcpy(p, cypher_text, cypher_text_len);
-	p += cypher_text_len;
-	tmplen += cypher_text_len;
+	r = sc_asn1_put_tag(0x82, cypher_text, cypher_text_len, p, output_len - (p - output_buf), &p);
+	if (r != SC_SUCCESS) {
+		goto err;
+	}
 
 	/* Sanity check the lengths again */
-	if(output_len != (size_t)tmplen) {
+	tmplen = sc_asn1_put_tag(0x7C, NULL, tmplen, NULL, 0, NULL)
+		+ sc_asn1_put_tag(0x82, NULL, cypher_text_len, NULL, 0, NULL);
```

Comment:
clang-format suggested the following change:


```
-	tmplen = sc_asn1_put_tag(0x7C, NULL, tmplen, NULL, 0, NULL)
-		+ sc_asn1_put_tag(0x82, NULL, cypher_text_len, NULL, 0, NULL);
+	tmplen = sc_asn1_put_tag(0x7C, NULL, tmplen, NULL, 0, NULL) + sc_asn1_put_tag(0x82, NULL, cypher_text_len, NULL, 0, NULL);
```

<!-- identifier: clang-format -->
<!-- File: src/libopensc/card-piv.c -->